src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -0,0 +1,164 @@
+
+
+def rpc_method(method: Callable[..., _T]) -> Callable[..., _T]:
+    method = abstractmethod(method)
+    setattr(method, _IS_RPC_METHOD_ATTR, True)
+    return method
+
+
+class RpcService(InterfaceWithLifecycle):
+    name: str
+    local_mode: SiloMode
+
+    def __init_subclass__(cls) -> None:
+        if cls._declares_service_interface():
+            # These class attributes are required on any RpcService subclass that has
+            # at least one method decorated by `@rpc_method`. (They can be left off
+            # if and when we make an intermediate abstract class.)
+            if not isinstance(getattr(cls, "name", None), str):
+                raise TypeError("`name` class attribute (str) is required")
+            if not isinstance(getattr(cls, "local_mode", None), SiloMode):
+                raise TypeError("`local_mode` class attribute (SiloMode) is required")
+
+    @classmethod
+    def _get_all_abstract_rpc_methods(cls) -> Iterator[Callable[..., Any]]:
+        for attr_name in dir(cls):
+            attr = getattr(cls, attr_name, None)
+            if callable(attr) and getattr(attr, _IS_RPC_METHOD_ATTR, False):
+                yield attr
+
+    @classmethod
+    def _declares_service_interface(cls) -> bool:
+        """Check whether a subclass declares the service interface.
+
+        By "service interface", we mean the set of RPC methods that are offered.
+        Those methods are decorated by `@rpc_method`, so we return true for any
+        service class that declares (but does not implement) at least one such method.
+
+        This method would return false, for example, for the local, database-backed
+        implementation that inherits from the base service. It also would return
+        false on an intermediate, abstract subclass that is meant to be extended to
+        declare other base services.
+        """
+
+        try:
+            next(cls._get_all_abstract_rpc_methods())
+        except StopIteration:
+            return False
+        else:
+            return True
+
+    @classmethod
+    @abstractmethod
+    def get_local_implementation(cls) -> "RpcService":
+        """Return a service object that runs locally.
+
+        The returned service object is (generally) the database-backed instance that
+        is called when we either receive a remote call from outside, or want to call
+        it within the same silo.
+
+        A base service class generally should override this class method, making a
+        forward reference to its own database-backed subclass.
+        """
+
+        raise NotImplementedError
+
+    @classmethod
+    def _create_local_delegation(cls) -> "RpcService":
+        """Create a lazy wrapper around the local implementation.
+
+        The indirection is necessary so that a base service's
+        `get_local_implementation` body can safely make a forward reference that
+        `resolve_to_delegation` won't call yet.
+        """
```

Comment:
This extra delegation step is admittedly pretty gnarly, but I think it's worth it to make the various subclasses simple and tidy. As long as it doesn't turn out to complicate things too much when looking through a stack trace or debugger.

Reply:
Aspects of this seem to be shared with `CreateStubFromBase` and `DelegatedBySiloMode`. In particular, I'm concerned about threading here.  I had to add thread locks around lazy delegation into the DelegatedBySiloMode because... sadly all of our acceptance tests use threaded servers in a shared process, so these things will get instantiated and run in non deterministic ordering.  The acceptance test loads the home page and sends off dozens of XHR requests to the threaded service all at once, and it can get messy.

Reply:
Yep, the similarity occurred to me and I meant to mention it here. If you're confident in the `DelegatedBySiloMode` implementation, I'll look into reusing it here.

We could also just jettison the lazy initialization and do something uglier with `resolve_to_delegation`, if necessary.

Reply:
Yeah, I'm ok with using a new implementation in for now, and then looking back at wether we can repurpose existing ones.  The only thing I think is strictly necessary is thread locking around instantiation.

Reply:
A related problem, as you can see from the current CI state, is that the auto-generated delegate subclasses aren't happy with `FilterQueryInterface` as a superclass.

Reply:
Let me take some time on this and ponder a bit more.

Reply:
ok cool!